pinot-core/src/test/java/org/apache/pinot/queries/HistogramQueriesTest.java

```diff
@@ -0,0 +1,251 @@
+    segmentGeneratorConfig.setTableName(RAW_TABLE_NAME);
+    segmentGeneratorConfig.setSegmentName(SEGMENT_NAME);
+    segmentGeneratorConfig.setOutDir(INDEX_DIR.getPath());
+
+    SegmentIndexCreationDriverImpl driver = new SegmentIndexCreationDriverImpl();
+    driver.init(segmentGeneratorConfig, new GenericRowRecordReader(records));
+    driver.build();
+
+    ImmutableSegment immutableSegment = ImmutableSegmentLoader.load(new File(INDEX_DIR, SEGMENT_NAME), ReadMode.mmap);
+    _indexSegment = immutableSegment;
+    _indexSegments = Arrays.asList(immutableSegment, immutableSegment);
+  }
+
+  @Test
+  public void testAggregationOnlyVer1() {
+    String query = "SELECT HISTOGRAM(intColumn,ARRAY[0,1,10,100,1000,10000]) FROM testTable";
+
+    // Inner segment
+    Object operator = getOperator(query);
+    assertTrue(operator instanceof AggregationOperator);
+    IntermediateResultsBlock resultsBlock = ((AggregationOperator) operator).nextBlock();
+    QueriesTestUtils.testInnerSegmentExecutionStatistics(((Operator) operator).getExecutionStatistics(), NUM_RECORDS, 0,
+        NUM_RECORDS, NUM_RECORDS);
+    List<Object> aggregationResult = resultsBlock.getAggregationResult();
+    assertNotNull(aggregationResult);
+    assertEquals(((DoubleArrayList) aggregationResult.get(0)).elements(), new double[]{1, 9, 90, 900, 1000});
+
+    query = "SELECT HISTOGRAM(intColumn,ARRAY[\"-Infinity\",1,10,100,1000,\"Infinity\"]) FROM testTable";
+
+    // Inner segment
+    operator = getOperator(query);
+    assertTrue(operator instanceof AggregationOperator);
+    resultsBlock = ((AggregationOperator) operator).nextBlock();
+    QueriesTestUtils.testInnerSegmentExecutionStatistics(((Operator) operator).getExecutionStatistics(), NUM_RECORDS, 0,
+        NUM_RECORDS, NUM_RECORDS);
+    aggregationResult = resultsBlock.getAggregationResult();
+    assertNotNull(aggregationResult);
+    assertEquals(((DoubleArrayList) aggregationResult.get(0)).elements(), new double[]{1, 9, 90, 900, 1000});
+
+    operator = getOperatorWithFilter(query);
+    assertTrue(operator instanceof AggregationOperator);
+    resultsBlock = ((AggregationOperator) operator).nextBlock();
+    QueriesTestUtils.testInnerSegmentExecutionStatistics(((Operator) operator).getExecutionStatistics(), 1500, 0, 1500,
+        NUM_RECORDS);
+    aggregationResult = resultsBlock.getAggregationResult();
+    assertNotNull(aggregationResult);
+    assertEquals(((DoubleArrayList) aggregationResult.get(0)).elements(), new double[]{0, 0, 0, 500, 1000});
+
+    // Inter segment
+    BrokerResponseNative brokerResponse = getBrokerResponse(query);
+    ResultTable resultTable = brokerResponse.getResultTable();
+    List<Object[]> rows = resultTable.getRows();
+    assertEquals(rows.get(0)[0], new double[]{4, 36, 360, 3600, 4000});
+  }
+
+  @Test
+  public void testAggregationOnlyVer2() {
+    String query = "SELECT HISTOGRAM(intColumn,0,1000,10) FROM testTable";
```

Comment:
Just for completeness of correctness testing, can we also try following ?

- HISTOGRAM(intColumn,0,100,9) - there is a leftover bin (if unsupported then verify that it throws error)
- HISTOGRAM(intColumn,0,17,1) - 17 bins (wanted to try a prime number) each of length 1
- HISTOGRAM(intColumn,0,17,17) - 1 bin of length 17
- For 2nd and 3rd above, let's make sure we have coverage for even, odd length bins as well
- HISTOGRAM(intColumn,0,1,1) - 1 bin of length 1
- HISTOGRAM(intColumn,5,25,5) - start point non-zero, 4 bins of length 5 each

Reply:
Added